x/mongo/driver/operation.go

```diff
@@ -1186,6 +1266,28 @@ func (op Operation) createMsgWireMessage(ctx context.Context, maxTimeMS uint64,
 	return bsoncore.UpdateLength(dst, wmindex, int32(len(dst[wmindex:]))), info, nil
 }
 
+// isLegacyHandshake returns True if the operation is the first message of
+// the initial handshake and should use a legacy hello.
+func isLegacyHandshake(op Operation, desc description.SelectedServer) bool {
+	isInitialHandshake := desc.WireVersion == nil || desc.WireVersion.Max == 0
+
+	return op.Legacy == LegacyHandshake && isInitialHandshake
+}
+
+func (op Operation) createWireMessage(
+	ctx context.Context,
+	dst []byte,
+	desc description.SelectedServer,
+	maxTimeMS uint64,
+	conn Connection,
+) ([]byte, startedInformation, error) {
+	if isLegacyHandshake(op, desc) {
```

Comment:
See [here](https://github.com/mongodb/specifications/blob/master/source/mongodb-handshake/handshake.rst#connection-handshake) for the exact logic of this: 

> If server API version is not requested and loadBalanced: False, drivers MUST use legacy hello for the first message of the initial handshake with the OP_QUERY protocol (before switching to OP_MSG if the maxWireVersion indicates compatibility), and include helloOk:true in the handshake request.